samples/compiler_plugins/xnnpack/src/xnnpack/Transforms/SetIdAttributes.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Visitors.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "xnnpack/IR/XnnpackOps.h"
+#include "xnnpack/Transforms/Passes.h"
+
+namespace mlir::iree_compiler::IREE::Xnnpack {
+#define GEN_PASS_DEF_SETIDATTRIBUTES
+#include "xnnpack/Transforms/Passes.h.inc"
+
+namespace {
+// Checks if `val` is a constant by going up the use-def tree to see if the
+// leaves of the tree are `stablehlo` constants.
+//
+// Note: a returned value of false means that this function was unable to
+// determine if the value is constant or not.
+static bool isConstant(Value val) {
+  SmallVector<Value> stack{val};
+  SmallVector<Operation *> zeroOperandOps;
+  while (!stack.empty()) {
+    Value currVal = stack.pop_back_val();
+    if (auto defOp = currVal.getDefiningOp()) {
+      if (defOp->getNumOperands() == 0) zeroOperandOps.push_back(defOp);
```

Comment:
Since the time I added the XOR changes, there have been improvements to int4 support in MLIR. I can now create an int4 constant directly without having to cast from i8->i4. This simplifies things.

Here is the fix addressing this comment: https://github.com/ramiro050/iree/pull/29/commits/d1276f45712672a7b3ecbbcecee1a4ebff8a3895